src/masternodes/mn_checks.cpp

```diff
@@ -3361,56 +3445,61 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
                 if (!rate)
                     return Res::Err("Cannot get interest rate for this token (%s)!", loanToken->symbol);
 
-                LogPrint(BCLog::LOAN,"CLoanPaybackLoanMessage()->%s->", loanToken->symbol); /* Continued */
-                auto subInterest = TotalInterest(*rate, height);
+                const auto scheme = mnview.GetLoanScheme(vault->schemeId);
+                if (!scheme) {
+                    return Res::Err("No such scheme id %s", vault->schemeId);
+                }
+
+                auto subInterest = TotalInterest(*rate, height, currentLoanAmount, loanToken->interest, scheme->rate);
                 auto subLoan = paybackAmount - subInterest;
 
                 if (paybackAmount < subInterest)
                 {
                     subInterest = paybackAmount;
                     subLoan = 0;
                 }
-                else if (it->second - subLoan < 0)
+                else if (currentLoanAmount - subLoan < 0)
                 {
-                    subLoan = it->second;
+                    subLoan = currentLoanAmount;
                 }
 
                 res = mnview.SubLoanToken(obj.vaultId, CTokenAmount{loanTokenId, subLoan});
                 if (!res)
                     return res;
 
-                LogPrint(BCLog::LOAN,"CLoanPaybackLoanMessage()->%s->", loanToken->symbol); /* Continued */
-                res = mnview.EraseInterest(height, obj.vaultId, vault->schemeId, loanTokenId, subLoan, subInterest);
+                res = mnview.EraseInterest(height, obj.vaultId, vault->schemeId, loanTokenId, subLoan, std::abs(subInterest));
                 if (!res)
                     return res;
 
-                if (static_cast<int>(height) >= consensus.FortCanningMuseumHeight && subLoan < it->second)
+                if (height >= static_cast<uint32_t>(consensus.FortCanningMuseumHeight) && subLoan < currentLoanAmount)
                 {
                     auto newRate = mnview.GetInterestRate(obj.vaultId, loanTokenId, height);
                     if (!newRate)
                         return Res::Err("Cannot get interest rate for this token (%s)!", loanToken->symbol);
 
-                    if (newRate->interestPerBlock == 0)
+                    if (height <= static_cast<uint32_t>(consensus.GreatWorldHeight) && newRate->interestPerBlock.amount == 0)
                         return Res::Err("Cannot payback this amount of loan for %s, either payback full amount or less than this amount!", loanToken->symbol);
                 }
 
                 CalculateOwnerRewards(obj.from);
 
                 if (paybackTokenId == loanTokenId)
                 {
-                    res = mnview.SubMintedTokens(loanTokenId, subLoan);
+                    // If interest was negative remove it from sub amount
+                    res = mnview.SubMintedTokens(loanTokenId, subInterest > 0 ? subLoan : subLoan + subInterest);
```

Comment:
again: negative interest does not reduce minted tokens. only the subLoan has to be substracted from minted tokens!